System/FilePath/Internal.hs

```diff
@@ -827,6 +828,49 @@ makeRelative root path
         takeAbs x | hasLeadingPathSeparator x && not (hasDrive x) = [pathSeparator]
         takeAbs x = map (\y -> if isPathSeparator y then pathSeparator else toLower y) $ takeDrive x
 
+-- | Strip the given directory from the filepath if and only if
+-- the given directory is a prefix of the filepath.
+--
+-- >>> stripFilePath "app" "app/File.hs"
+-- Just "File.hs"
+--
+-- >>> stripFilePath "src" "app/File.hs"
+-- Nothing
+--
+-- >>> stripFilePath "src" "src-dir/File.hs"
+-- Nothing
+--
+-- >>> stripFilePath "." "src/File.hs"
+-- Just "src/File.hs"
+--
+-- >>> stripFilePath "app/" "./app/Lib/File.hs"
+-- Just "Lib/File.hs"
+--
+-- >>> stripFilePath "/app/" "./app/Lib/File.hs"
+-- Nothing -- Nothing since '/app/' is absolute
+--
+-- >>> stripFilePath "/app" "/app/Lib/File.hs"
+-- Just "Lib/File.hs"
+stripFilePath :: FilePath -> FilePath -> Maybe FilePath
+stripFilePath "." fp
+  | isRelative fp = Just fp
+  | otherwise = Nothing
+stripFilePath dir' fp'
+  | Just relativeFpParts <- splitDir `stripPrefix` splitFp = Just (joinPath relativeFpParts)
+  | otherwise = Nothing
+  where
+    dir = normalise dir'
```

Comment:
I'm a friend of `ViewPatterns` for this, unless there's a reason we can't use that here

Reply:
I think the whole implementation is a bit trash atm, so no reason other than trying to be as simple as possible